src/main.rs

```diff
@@ -1,3 +1,204 @@
+        }
+    }
+
+    async fn run_linting_and_report_diagnostics(&self, uri: &Url) {
+        let per_file_state = self.per_file.get(uri).unwrap();
+        let violations = tree_sitter_lint_local::run_for_slice(
+            &per_file_state.contents,
+            Some(&per_file_state.tree),
+            "dummy_path",
+            Default::default(),
+        );
+        self.client
+            .publish_diagnostics(
+                uri.clone(),
+                violations
+                    .into_iter()
+                    .map(|violation| Diagnostic {
+                        message: violation.message,
+                        range: tree_sitter_range_to_lsp_range(
+                            &per_file_state.contents,
+                            violation.range,
+                        ),
+                        severity: Some(DiagnosticSeverity::ERROR),
+                        code: Some(NumberOrString::String(violation.rule.name)),
+                        source: Some("tree-sitter-lint".to_owned()),
+                        ..Default::default()
+                    })
+                    .collect(),
+                None,
+            )
+            .await;
+    }
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for Backend {
+    async fn initialize(&self, _: InitializeParams) -> Result<InitializeResult> {
+        Ok(InitializeResult {
+            capabilities: ServerCapabilities {
+                text_document_sync: Some(TextDocumentSyncCapability::Kind(
+                    TextDocumentSyncKind::INCREMENTAL,
+                )),
+                ..Default::default()
+            },
+            ..Default::default()
+        })
+    }
+
+    async fn initialized(&self, _: InitializedParams) {
+        // self.client
+        //     .log_message(tower_lsp::lsp_types::MessageType::INFO, "server initialized!")
+        //     .await;
+    }
+
+    async fn shutdown(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        let contents: Rope = (&*params.text_document.text).into();
+        self.per_file.insert(
+            params.text_document.uri.clone(),
+            PerFileState {
+                tree: parse_from_scratch(&contents),
+                contents,
+            },
+        );
+
+        self.run_linting_and_report_diagnostics(&params.text_document.uri)
+            .await;
+    }
+
+    async fn did_change(&self, params: DidChangeTextDocumentParams) {
+        {
+            let mut file_state = self
+                .per_file
+                .get_mut(&params.text_document.uri)
+                .expect("Changed file wasn't loaded");
+            for content_change in &params.content_changes {
+                match content_change.range {
+                    Some(range) => {
+                        let start_char =
+                            lsp_position_to_char_offset(&file_state.contents, range.start);
```

Comment:
A lot of these offset-conversion routines are questionable

That one "sober" post was saying something about LSP using UTF-16 character offsets and it being hard to get right

I'm guessing as of now "it works fine if your whole file is simple ASCII characters" but probably gets inaccurate if not